### ISSUE TITLE
Nodes with different feature size

### ISSUE THREAD

**Issue by ArashJavan** [pyg-team/pytorch_geometric#1950]
## ❓ Questions & Help
Hi,
let say I have a simple undirected graph made by only three nodes, where each node has a different feature-length.

![grpah](https://user-images.githubusercontent.com/5918827/103350048-00496e80-4a9f-11eb-869b-4ab0697ca2da.png)

How can I apply the Graph Neural Network on this graph with diff. feature-length? I know one way could be to just
pad them all to a specific length by just filling them with zeros. But is it also possible without doing this? When yes on which graph (or paper) is it done?

Thanks and happy new year in advance!

**Comment by rusty1s** [comment]
In general, this is only possible by representing the graph as a heterogeneous one. For example, you can look at [these example](https://github.com/snap-stanford/ogb/tree/master/examples/nodeproppred/mag) from the OGB suite on how to handle heterogeneous graphs.

Even if you pad them with zeros, the meaning of individual features might be different. However, in your case, you might be able to represent your features as a 0/1 encoding, e.g.
```python
x = [
   [0, 0, 1, 0, 0, 0, 0, 1, 1],
   [1, 1, 1, 1, 0, 0, 0, 0, 0],
   [1, 0, 0, 0, 1, 0, 0, 0, 0],
] 
```

**Comment by ArashJavan** [comment]
@rusty1s thanks for your reply. Well, the features are continuous (float) - the above picture was just an example :) 
So that means the only possible solution would be using heterogeneous graphs. right! As far as I can see the only example using heterogeneous graphs in the link you sent, is the [R-GCN](https://github.com/snap-stanford/ogb/blob/master/examples/nodeproppred/mag/rgcn.py), right?

I am pretty new in the field of GCNN, but it seems to me that, DGL has a specific class for hetero-graphs. Does py-geometric also offer something similar?

Nevertheless, I am working on time-series, where each node has a different number of features, so temporal info is also importent. Do you know any paper/implementation, handling this kind of networks? I could not find any!

Thanks

**Comment by reshinthadithyan** [comment]
@rusty1s why don't we provide mask vectors like huggingface/transformers do ? By doing this we can solve a lot of problem while storing the data. I'm not sure how we'll handle the feature vector.
In the given example, we can pad and use that mask vector to identify the padded values.
`x0 = [3,8,9,0] mask = [1,1,1,0],     
 x1 = [4,3,2,1] mask = [1,1,1,1],  
x2 = [1,5,0,0] mask = [1,1,0,0]`

**Comment by rusty1s** [comment]
Well, one can of course pad the input features via `torch.pad_sequence`, but this does not make much sense in the graph context where you expect features to have the same meaning across the feature dimension. If one wants to use sequences as input, one can always use an LSTM/Transformer as feature encoder, and input that into a GNN.

Depending on your problem, it might make sense to treat this graph as a hetero one (where you have different node and edge types with different message passing operators). It's true that DGL has a dedicated class for hetero graphs, while PyG does not. PyG expects hetero graphs to be stored in dictionaries:
```python
data.edge_index_dict = {
  ('paper', 'cites', 'paper'): ... [2. num_edges1]
  ('author', 'writes', 'paper'), ... [2, num_edges2]
}
data.x_dict = {
  'paper': ... [num_nodes1, num_features1]
  'author': ... [num_nodes2], num_features2]
```
from which you then can simply call different GNN ops:
```python
conv1 = SAGEConv((num_features2, num_features1), 256)
conv2 = SAGEConv(num_features1, 256)

out = conv1((data.x_dict['author'], data.x_dict['paper']), data.edge_index_dict[('author', 'writes', 'paper')])
out += conv2(data.x_dict['paper'], data.edge_index_dict[('paper', 'cites', 'paper')])
```
